src/overlay/Peer.cpp

```diff
@@ -630,38 +710,256 @@ Peer::recvMessage(StellarMessage const& stellarMsg)
+                    FMT_STRING("Error RecvMessage T:{} cat:{} {} @{:d}"), mtype,
+                    cat, self->toString(), app.getConfig().PEER_PORT);
+                CLOG_ERROR(Overlay, "Dropping connection with {}: {}", err,
+                           e.what());
+                self->drop("Bad crypto request",
+                           Peer::DropDirection::WE_DROPPED_REMOTE,
+                           Peer::DropMode::IGNORE_WRITE_QUEUE);
             }
+
+            // Notify on completion, schedule more reading if necessary
+            self->messageProcessed(sm);
         },
         fmt::format(FMT_STRING("{} recvMessage"), cat), type);
 }
 
+void
+Peer::sendSendMore(uint32_t numMessages)
+{
+    ZoneScoped;
+    StellarMessage m;
+    m.type(SEND_MORE);
+    m.sendMoreMessage().numMessages = numMessages;
+    sendMessage(m);
+}
+
+void
+Peer::recvSendMore(StellarMessage const& msg)
+{
+    ZoneScoped;
+    // Flow control must be "on"
+    assert(mRemoteOverlayVersion >=
+           Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL);
+    assert(mRemotePeerWantsFlowControl && *mRemotePeerWantsFlowControl);
+    assert(msg.sendMoreMessage().numMessages > 0);
+
+    CLOG_TRACE(Overlay, "Peer {} sent SEND_MORE {}",
+               mApp.getConfig().toShortString(getPeerID()),
+               msg.sendMoreMessage().numMessages);
+
+    // SEND_MORE means we can free some capacity, and dump the next batch of
+    // messages onto the writing queue
+    mOutboundCapacity += msg.sendMoreMessage().numMessages;
+    maybeShedLoadAndSendNextBatch();
+}
+
+bool
+Peer::hasReadingCapacity() const
+{
+    bool enabledAndSupportedByRemotePeer =
+        (mRemoteOverlayVersion >= Peer::FIRST_VERSION_SUPPORTING_FLOW_CONTROL &&
+         mApp.getConfig().ENABLE_OVERLAY_FLOW_CONTROL);
+    return !enabledAndSupportedByRemotePeer || mCapacity.mTotalCapacity > 0;
+}
+
+static medida::Timer&
+getOutboundFloodTimer(Application& app, StellarMessage const& msg)
+{
+    std::string type;
+    switch (msg.type())
+    {
+    case SCP_MESSAGE:
+        type = "scp";
+        break;
+    case TRANSACTION:
+        type = "tx";
+        break;
+    default:
+        type = "ctrl";
+    }
+
+    return app.getMetrics().NewTimer({"overlay", "outbound-queue", type});
+}
+
+void
+Peer::maybeTrimQueue()
+{
+    auto now = mApp.getClock().now();
+    auto shouldTrim = [&](QueuedOutboundMessage const& msg) -> bool {
+        if (msg.mMessage.type() == TRANSACTION)
+        {
+            // Trim transactions that sit in the queue for too long
+            return (now - msg.mTimeEmplaced) >
+                   Peer::OUTBOUND_TRANSACTION_DROP_TIMEOUT;
+        }
+        else if (msg.mMessage.type() == SCP_MESSAGE)
+        {
+            auto slotIndex = msg.mMessage.envelope().statement.slotIndex;
+            auto minSlotToRemember =
+                mApp.getHerder().getMinLedgerSeqToRemember();
+            if (slotIndex >= minSlotToRemember &&
```

Comment:
maybe you can avoid scanning all SCP messages for all slots if you just pass as a "hint" to this trim function the SCP message that you just added to the queue, that way you can focus on messages for a specific validator.

This hint could also be used to avoid trimming SCP messages when you send a transaction

Reply:
This may actually be better than throttling as you will perform a lot less comparisons.